src/node.cc

```diff
@@ -3828,11 +3828,9 @@ static void StartNodeInstance(void* arg) {
       if (more == false) {
         EmitBeforeExit(env);
 
-        // Emit `beforeExit` if the loop became alive either after emitting
-        // event, or after running some callbacks.
-        more = uv_loop_alive(env->event_loop());
-        if (uv_run(env->event_loop(), UV_RUN_NOWAIT) != 0)
-          more = true;
+        more = env->event_loop()->active_handles ||
+            (env->event_loop()->active_reqs[0] !=
+             env->event_loop()->active_reqs[1]);
```

Comment:
You're probably well aware of this but active_handles and active_reqs are implementation details.